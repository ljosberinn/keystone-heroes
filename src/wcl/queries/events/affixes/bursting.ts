import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const BURSTING = {
  damage: 243_237,
  debuff: 240_443,
};

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=4&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20243237%20and%20rawDamage%20%3E%200
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 5987472, endTime: 7736379, limit: 2000, filterExpression: "type = \"damage\" and target.type = \"player\" and ability.id = 243237 and rawDamage > 0") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "damage" and target.type = "player" and ability.id = ${BURSTING.damage} and rawDamage > 0`,
  `type in ("applydebuff", "applydebuffstack") and ability.id = ${BURSTING.debuff}`,
];

const isBurstingDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: BURSTING.damage,
});

const isBurstingApplyDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: BURSTING.debuff,
});

const isBurstingApplyDebuffStackEvent =
  createIsSpecificEvent<ApplyDebuffStackEvent>({
    type: "applydebuffstack",
    abilityGameID: BURSTING.debuff,
  });

export const getBurstingEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffEvent | ApplyDebuffStackEvent)[] => {
  if (!affixSet.has(Affixes.Bursting)) {
    return [];
  }

  return [
    ...allEvents.filter(isBurstingDamageEvent),
    ...allEvents.filter(isBurstingApplyDebuffEvent),
    ...allEvents.filter(isBurstingApplyDebuffStackEvent),
  ];
};
