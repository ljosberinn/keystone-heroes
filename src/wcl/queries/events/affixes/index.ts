import { Affixes } from "@prisma/client";

import type { EventParams } from "..";
import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
  RemoveBuffEvent,
  RemoveDebuffEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
  | "Infernal"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
  [Affixes.Encrypted]: [],
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): {
  explosiveTargetID: number | null;
  events: (
    | ApplyBuffEvent
    | ApplyBuffStackEvent
    | DamageEvent
    | HealEvent
    | InterruptEvent
    | ApplyDebuffStackEvent
    | BeginCastEvent
    | CastEvent
    | ApplyDebuffEvent
    | RemoveBuffEvent
    | RemoveDebuffEvent
  )[];
} => {
  const affixSet = new Set(affixes);

  const { explosiveTargetID, events: explosiveEvents } = getExplosiveEvents(
    allEvents,
    affixSet
  );

  return {
    explosiveTargetID,
    events: [
      // seasonal
      ...getTormentedEvents(allEvents, affixSet),
      // common affixes
      ...getStormingEvents(allEvents, affixSet),
      ...getSpitefulEvents(allEvents, affixSet),
      ...getSanguineEvents(allEvents, affixSet),
      ...getVolcanicEvents(allEvents, affixSet),
      ...getQuakingEvents(allEvents, affixSet),
      ...getBolsteringEvents(allEvents, affixSet),
      ...getBurstingEvents(allEvents, affixSet),
      ...explosiveEvents,
      ...getGrievousEvents(allEvents, affixSet),
      ...getNecroticEvents(allEvents, affixSet),
    ],
  };
};
