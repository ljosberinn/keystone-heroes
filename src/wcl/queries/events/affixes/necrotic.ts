import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const NECROTIC = 209_858;

/**
 * @see https://www.warcraftlogs.com/reports/DmVgxdj3WTYBApQt/#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24ability.id%20%3D%20209858%20and%20type%20%3D%20%22applydebuffstack%22%20and%20stack%20%3E%2010%5E2%24Off%24%23909049%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20209858%20and%20rawDamage%20%3E%200
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "DmVgxdj3WTYBApQt") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 0, endTime: 1414930, filterExpression: "(target.type = \"player\" and ability.id = 209858) and ((type = \"applydebuffstack\" and stack > 10) or (type = \"damage\" and rawDamage > 0))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */

export const filterExpression = [
  `type in ("applydebuff", "damage", "applydebuffstack") and ability.id = ${NECROTIC}`,
];

const isNecroticDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NECROTIC,
});

const isNecroticDebuffStackEvent = createIsSpecificEvent<ApplyDebuffStackEvent>(
  {
    type: "applydebuffstack",
    abilityGameID: NECROTIC,
  }
);

const isNecroticDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: NECROTIC,
});

export const getNecroticEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffStackEvent | ApplyDebuffEvent)[] => {
  if (!affixSet.has(Affixes.Necrotic)) {
    return [];
  }

  return [
    ...allEvents.filter(isNecroticDebuffStackEvent),
    ...allEvents.filter(isNecroticDamageEvent),
    ...allEvents.filter(isNecroticDebuffEvent),
  ];
};
