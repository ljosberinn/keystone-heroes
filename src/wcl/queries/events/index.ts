import { Affixes } from "@prisma/client";

import type { DungeonIDs } from "../../../db/data/dungeons";
import type {
  AllTrackedEventTypes,
  BeginCastEvent,
  DamageEvent,
  DeathEvent,
} from "../events/types";
import { filterAffixEvents, getAffixExpression } from "./affixes";
import { findExplosiveTargetID } from "./affixes/explosive";
import {
  cheatDeathFilterExpression,
  filterCheatDeathEvents,
} from "./cheathDeath";
import {
  generalCovenantExpression,
  filterCovenantCastEvents,
} from "./covenant";
import {
  deathFilterExpression,
  filterEnemyDeathEvents,
  filterPlayerDeathEvents,
} from "./death";
import { getDungeonExpression, filterDungeonEvents } from "./dungeons";
import {
  filterPlayerInterruptEvents,
  interruptFilterExpression,
} from "./interrupts";
import {
  engineeringBattleRezExpression,
  invisibilityFilterExpression,
  filterProfessionEvents,
} from "./professions";
import {
  filterRemarkableSpellEvents,
  remarkableSpellFilterExpression,
} from "./spells";
import { filterTrinkets, trinketsFilterExpression } from "./trinkets";
import { recursiveGetEvents } from "./utils";

export type EventParams = {
  reportID: string;
  startTime: number;
  endTime: number;
  dungeonID: DungeonIDs;
  fightID: number;
  affixes: Affixes[];
};

const generateFilterExpression = ({
  dungeonID,
  affixes,
}: Pick<EventParams, "dungeonID" | "affixes">) => {
  return [
    deathFilterExpression,
    remarkableSpellFilterExpression,
    engineeringBattleRezExpression,
    invisibilityFilterExpression,
    interruptFilterExpression,
    generalCovenantExpression,
    cheatDeathFilterExpression,
    trinketsFilterExpression,
    ...getDungeonExpression(dungeonID),
    ...getAffixExpression(affixes),
  ]
    .flat()
    .map((part) => `(${part})`)
    .join(" or ");
};

export const getEvents = async (
  params: EventParams,
  playerMetaInformation: { actorID: number; class: number }[]
): Promise<{
  allEvents: AllTrackedEventTypes[];
  playerDeathEvents: DeathEvent[];
  enemyDeathEvents: (DeathEvent | BeginCastEvent | DamageEvent)[];
  explosiveTargetID: number | null;
}> => {
  const filterExpression = generateFilterExpression({
    dungeonID: params.dungeonID,
    affixes: params.affixes,
  });

  const allEvents = await recursiveGetEvents<AllTrackedEventTypes>({
    startTime: params.startTime,
    endTime: params.endTime,
    reportID: params.reportID,
    filterExpression,
  });

  const actorIDSet = new Set(
    playerMetaInformation.map((dataset) => dataset.actorID)
  );

  const dungeonEvents = filterDungeonEvents(
    allEvents,
    params.dungeonID,
    actorIDSet
  );
  const { explosiveTargetID, events: affixEvents } = filterAffixEvents(
    allEvents,
    params.affixes
  );
  const professionEvents = filterProfessionEvents(allEvents);
  const remarkableSpellEvents = filterRemarkableSpellEvents(allEvents);
  const playerDeathEvents = filterPlayerDeathEvents(
    allEvents,
    actorIDSet
    // remarkableSpellEvents
  );
  const interruptEvents = filterPlayerInterruptEvents(allEvents);
  const enemyDeathEvents = filterEnemyDeathEvents(
    allEvents,
    actorIDSet,
    params.affixes.includes(Affixes.Explosive)
      ? findExplosiveTargetID(allEvents)
      : null
  );
  const sharedCovenantAbilitiesEvents = filterCovenantCastEvents(allEvents);
  const cheatDeathEvents = filterCheatDeathEvents(allEvents);
  const trinketEvents = filterTrinkets(allEvents);

  return {
    allEvents: [
      ...professionEvents,
      ...dungeonEvents,
      ...affixEvents,
      ...remarkableSpellEvents,
      ...playerDeathEvents,
      ...interruptEvents,
      ...sharedCovenantAbilitiesEvents,
      ...cheatDeathEvents,
      ...trinketEvents,
    ].sort((a, b) => a.timestamp - b.timestamp),
    playerDeathEvents,
    enemyDeathEvents,
    explosiveTargetID,
  };
};
