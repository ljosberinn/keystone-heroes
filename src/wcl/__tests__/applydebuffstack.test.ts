import { NECROTIC } from "../queries/events/affixes/necrotic";
import type { ApplyDebuffStackEvent } from "../queries/events/types";
import { applyDebuffStackprocessor } from "../transform/events/applydebuffstack";
import allEvents from "./fixtures/allEvents.json";

describe("applydebuffstack", () => {
  const params = {
    targetPlayerID: 1,
    targetNPCID: null,
    sourcePlayerID: null,
    sourceNPCID: null,
  };

  const applydebuffstackEvents = allEvents.filter(
    (event): event is ApplyDebuffStackEvent => event.type === "applydebuffstack"
  );

  test("tracks Necrotic stacks", () => {
    const result = applydebuffstackEvents
      .filter((event) => event.abilityGameID === NECROTIC)
      .map((event) => applyDebuffStackprocessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("returns null on unknown applydebuffstack events", () => {
    expect(
      applyDebuffStackprocessor(
        {
          type: "applydebuffstack",
          abilityGameID: 1,
          sourceID: 1,
          stack: 1,
          targetID: 1,
          timestamp: 1,
        },
        params
      )
    ).toBeNull();
  });
});
