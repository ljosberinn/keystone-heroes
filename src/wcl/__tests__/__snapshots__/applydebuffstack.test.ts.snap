// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applydebuffstack tracks Necrotic stacks 1`] = `
[
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 23360,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 24411,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 24929,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 25626,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 15,
    "targetPlayerID": 1,
    "timestamp": 26719,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 16,
    "targetPlayerID": 1,
    "timestamp": 26774,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 17,
    "targetPlayerID": 1,
    "timestamp": 27277,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 18,
    "targetPlayerID": 1,
    "timestamp": 28057,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 19,
    "targetPlayerID": 1,
    "timestamp": 29928,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 20,
    "targetPlayerID": 1,
    "timestamp": 30143,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 21,
    "targetPlayerID": 1,
    "timestamp": 31422,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 22,
    "targetPlayerID": 1,
    "timestamp": 31846,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 23,
    "targetPlayerID": 1,
    "timestamp": 32492,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 78747,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 78997,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 79130,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 79876,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 15,
    "targetPlayerID": 1,
    "timestamp": 81102,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 16,
    "targetPlayerID": 1,
    "timestamp": 81102,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 17,
    "targetPlayerID": 1,
    "timestamp": 81347,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 18,
    "targetPlayerID": 1,
    "timestamp": 81493,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 19,
    "targetPlayerID": 1,
    "timestamp": 82231,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 20,
    "targetPlayerID": 1,
    "timestamp": 82231,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 21,
    "targetPlayerID": 1,
    "timestamp": 83450,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 22,
    "targetPlayerID": 1,
    "timestamp": 83705,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 23,
    "targetPlayerID": 1,
    "timestamp": 84592,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 24,
    "targetPlayerID": 1,
    "timestamp": 85806,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 25,
    "targetPlayerID": 1,
    "timestamp": 86061,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 26,
    "targetPlayerID": 1,
    "timestamp": 86198,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 27,
    "targetPlayerID": 1,
    "timestamp": 86628,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 28,
    "targetPlayerID": 1,
    "timestamp": 88550,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 29,
    "targetPlayerID": 1,
    "timestamp": 88563,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 30,
    "targetPlayerID": 1,
    "timestamp": 90547,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 31,
    "targetPlayerID": 1,
    "timestamp": 90561,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 32,
    "targetPlayerID": 1,
    "timestamp": 91939,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 33,
    "targetPlayerID": 1,
    "timestamp": 92091,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 34,
    "targetPlayerID": 1,
    "timestamp": 92563,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 35,
    "targetPlayerID": 1,
    "timestamp": 99337,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 36,
    "targetPlayerID": 1,
    "timestamp": 99828,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 37,
    "targetPlayerID": 1,
    "timestamp": 101684,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 171812,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 173313,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 174172,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 179731,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 15,
    "targetPlayerID": 1,
    "timestamp": 183796,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 16,
    "targetPlayerID": 1,
    "timestamp": 187171,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 273330,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 273770,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 276134,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 278493,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 337936,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 338343,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 339156,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 339269,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 15,
    "targetPlayerID": 1,
    "timestamp": 339269,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 16,
    "targetPlayerID": 1,
    "timestamp": 340365,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 17,
    "targetPlayerID": 1,
    "timestamp": 341618,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 18,
    "targetPlayerID": 1,
    "timestamp": 341979,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 19,
    "targetPlayerID": 1,
    "timestamp": 342026,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 20,
    "targetPlayerID": 1,
    "timestamp": 342095,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 21,
    "targetPlayerID": 1,
    "timestamp": 342336,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 22,
    "targetPlayerID": 1,
    "timestamp": 342351,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 23,
    "targetPlayerID": 1,
    "timestamp": 342367,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 24,
    "targetPlayerID": 1,
    "timestamp": 343641,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 25,
    "targetPlayerID": 1,
    "timestamp": 343979,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 26,
    "targetPlayerID": 1,
    "timestamp": 343996,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 27,
    "targetPlayerID": 1,
    "timestamp": 344355,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 28,
    "targetPlayerID": 1,
    "timestamp": 344392,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 29,
    "targetPlayerID": 1,
    "timestamp": 344687,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 30,
    "targetPlayerID": 1,
    "timestamp": 344687,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 31,
    "targetPlayerID": 1,
    "timestamp": 344687,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 32,
    "targetPlayerID": 1,
    "timestamp": 344705,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 33,
    "targetPlayerID": 1,
    "timestamp": 345180,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 349098,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 349403,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 349403,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 349536,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 15,
    "targetPlayerID": 1,
    "timestamp": 350185,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 16,
    "targetPlayerID": 1,
    "timestamp": 350772,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 17,
    "targetPlayerID": 1,
    "timestamp": 351056,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 18,
    "targetPlayerID": 1,
    "timestamp": 351776,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 19,
    "targetPlayerID": 1,
    "timestamp": 351889,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 20,
    "targetPlayerID": 1,
    "timestamp": 352039,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 21,
    "targetPlayerID": 1,
    "timestamp": 352039,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 22,
    "targetPlayerID": 1,
    "timestamp": 352531,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 23,
    "targetPlayerID": 1,
    "timestamp": 352546,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 425625,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 427414,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 427994,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 430343,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 15,
    "targetPlayerID": 1,
    "timestamp": 432704,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 11,
    "targetPlayerID": 1,
    "timestamp": 495488,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 12,
    "targetPlayerID": 1,
    "timestamp": 502908,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 13,
    "targetPlayerID": 1,
    "timestamp": 503134,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 14,
    "targetPlayerID": 1,
    "timestamp": 504092,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 15,
    "targetPlayerID": 1,
    "timestamp": 505932,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 16,
    "targetPlayerID": 1,
    "timestamp": 505962,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 17,
    "targetPlayerID": 1,
    "timestamp": 506463,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 18,
    "targetPlayerID": 1,
    "timestamp": 507980,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 19,
    "targetPlayerID": 1,
    "timestamp": 508525,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 20,
    "targetPlayerID": 1,
    "timestamp": 509124,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 21,
    "targetPlayerID": 1,
    "timestamp": 509799,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 22,
    "targetPlayerID": 1,
    "timestamp": 509827,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 23,
    "targetPlayerID": 1,
    "timestamp": 510339,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 24,
    "targetPlayerID": 1,
    "timestamp": 510873,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 25,
    "targetPlayerID": 1,
    "timestamp": 510946,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 26,
    "targetPlayerID": 1,
    "timestamp": 511549,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 27,
    "targetPlayerID": 1,
    "timestamp": 512144,
  },
  {
    "abilityID": 209858,
    "eventType": "ApplyDebuffStack",
    "stacks": 28,
    "targetPlayerID": 1,
    "timestamp": 512687,
  },
]
`;
